Reformatting and extending

Jos-Ven opened the pull request:
Added SendStk and sendtcp so a web server can pass messages and stacks to another web server for remote debugging. That also needed the chainperform.  
That leaded to the discovery to  a number of bugs (sorry). These update solves them.
Also increased the maximal number of calls to the OS and made f>d number base independent.
Extended the structure pack and added x:structures to the vocabulary environment
Then I decided to change the format of my submitted sources so that they are in line
with your standard and renamed the files that had the .f extension to the .fth extension.
All in all a lot of changes. Sorry for the inconveniences.

Jos-Ven committed:
Adding the spi interface

Jos-Ven committed:
Merge branch 'master' of https://github.com/Jos-Ven/cforth
For the spi interface

Jos-Ven committed:
remove interface.bak

Jos-Ven committed:
Merge branch 'MitchBradley:master' into master

Jos-Ven committed:
Solved the w/o problem and added sdcard.fth

Commit:
Solved the w/o problem and added sdcard.fth

Commit:
Solved the w/o problem and added sdcard.fth

Jos-Ven committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Jos-Ven committed:
corr

Commit:
To make a SVG-plot from data on a sdcard instead of memory

Commit:
Reformatted app.fth, solved a bug in ms and added .path

Commit:
Added setvbuf for a better performance on a sdcard

Commit:
Increasing the accuracy of f>d

Commit:
The updated plotter

Commit:
Adapted f>s so rounding follows the ANSI standard.

Jos-Ven committed:
Merge branch 'master' into master

Commit:
Extending and improving various tools

Commit:
Increased the maximal number of calls in extend.c to 160

Commit:
Made f>d independed of the number base

Commit:
Reformatting source text

Commit:
Bugfixes reformatting and extending

Commit:
Better handling of timedrift from deep sleep and reformatting

Commit:
Reformatting

Commit:
Corr.

Commit:
Renaming files conform the standard

MitchBradley commented on src/app/esp32-extra/app.fth:
```diff
@@ -27,38 +27,35 @@ alias m-init noop
-      then
+      #240000000 =  if  exit  else  #1 lshift  then
    then  #3 /
 ;
```
This is really hard to read.  Try a case statement

Jos-Ven replied:
done

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -1,26 +1,43 @@
+         then
+   else
+       0
+   then
```
Use false instead of 0 if the result is a flag.  Conveys the intent better.

Jos-Ven replied:
Done on my site, if not right then I will submit the file again.

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -32,66 +49,75 @@ alias init-sysled noop
+   >r
+   <#                         \ every 'seperator' digits from right
+   r@ 0 do # 2dup d0= ?leave loop
+        begin   2dup d0= 0=   \ while not a double zero
```
Why is the begin block indented by 7 instead of 3?

Jos-Ven replied:
corrected

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -104,460 +130,470 @@ test-1second
-       then
-    (time-start) get-system-time! ;
-
+    trim-time f0=  if
```
3 spaces before trim-time, not 4

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -104,460 +130,470 @@ test-1second
-
-: .elapsed ( - )
-   usf@  (time-start) 2@ us-elapsed fus>fsec fe. ." sec." ;
+     system-time>f f- trim-time f-
```
3 spaces, not 5

Jos-Ven replied:
corrected

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -104,460 +130,470 @@ test-1second
+   repeat drop saved-rp @ rp!
+;
+: last-xt, ( - )
+   s" lastacf  [ifndef]  id:  xtliteral  [else]  literal  [then] " evaluate ;
```
; on next line

Jos-Ven replied:
Is corrected on my site. If not then I will submit that file again.

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -104,460 +130,470 @@ test-1second
-2 value wifi-#retries
-0 value wifi-storage
--2 value wifi-logon-state   \ -2 not tried   -1 logon failed    0 Logon OK
+       ." Enter Password:" pad dup 50 accept r> place
```
indentation wrong

Jos-Ven replied:
Done on my site, if not right then I will submit the file again.

MitchBradley commented on src/app/esp32-extra/tools/extra.fth:
```diff
@@ -104,460 +130,470 @@ test-1second
-
+    ipaddr@ @ 0=  if
+         100 ms cr  ." No connection. Entering sleep mode..."  DeepSleep
+      else
```
indentation wrong

Jos-Ven replied:
Thank you for looking so close to my attempts to standardize the sources.
Tomorrow I will update the involved files.

MitchBradley reviewed (changes requested):
I didn't check everything but I did notice a lot of formatting discrepancies.

quozl commented:
I've looked at the changes.  Those to esp32-extra are not concerning, as I don't use that part of the source tree.  The changes to `src/cforth` are a bit more important to me.  I'm not sure what the impact of `f>d` and `f>s` changes will be.  It took a while to find the commit message, as the branch seems to have diverged and there are multiple merge commits.

Jos-Ven commented:
> I've looked at the changes. Those to esp32-extra are not concerning, as I don't use that part of the source tree. The changes to `src/cforth` are a bit more important to me. I'm not sure what the impact of `f>d` and `f>s` changes will be. It took a while to find the commit message, as the branch seems to have diverged and there are multiple merge commits.

At some time 'int' returned -2147483648 and f>d d>s did not.
Can't reproduce this now.

Commit:
Resolving formatting discrepancies and git conflicts.

Jos-Ven commented:
ms>ticks has also been changed, so it uses case. I hope it is alright now.

quozl commented:
Wondered why GitHub was noting conflicts.  Looking at the history using gitk, you're branch diverged from master just before 353a969f07027361f44b3ff9c5b7059be5780274 so you could either rebase on that and resolve the conflict markers, or cherry-pick and resolve.

Commit:
Cleanup

Jos-Ven commented:
> Wondered why GitHub was noting conflicts. Looking at the history using gitk, you're branch diverged from master just before [353a969](https://github.com/MitchBradley/cforth/commit/353a969f07027361f44b3ff9c5b7059be5780274) so you could either rebase on that and resolve the conflict markers, or cherry-pick and resolve.


Thank you for your help. I am afraid it is a bit more complicated.
My own site is OK here is what happened:

1)
0.pi@Acer~/cforth$ git pull --rebase
error: cannot pull with rebase: You have unstaged changes.
error: please commit or stash them.
2)
0.pi@Acer~/cforth$ git pull --rebase --autostash
Already up to date.

So I did a git pull --rebase form Mitch's site.

That got me:
0.pi@Acer~/cforth$ git pull --rebase https://github.com/MitchBradley/cforth
remote: Enumerating objects: 1, done.
remote: Counting objects: 100% (1/1), done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Unpacking objects: 100% (1/1), 1.10 KiB | 1.10 MiB/s, done.
From https://github.com/MitchBradley/cforth
 * branch            HEAD       -> FETCH_HEAD
Auto-merging build/esp32-extra/sdk_build/main/interface.c
CONFLICT (content): Merge conflict in build/esp32-extra/sdk_build/main/interface.c
Auto-merging src/app/esp32-extra/app.fth
Auto-merging src/app/esp32-extra/extend.c
CONFLICT (content): Merge conflict in src/app/esp32-extra/extend.c
Auto-merging src/app/esp32-extra/files.fth
CONFLICT (content): Merge conflict in src/app/esp32-extra/files.fth
Auto-merging src/app/esp32-extra/interface.h
Auto-merging src/app/esp32-extra/sdcard.fth
CONFLICT (add/add): Merge conflict in src/app/esp32-extra/sdcard.fth
Auto-merging src/app/esp32-extra/tools/extra.fth
CONFLICT (content): Merge conflict in src/app/esp32-extra/tools/extra.fth
error: could not apply 48a7eeb...  Solved the w/o problem and added sdcard.fth
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 48a7eeb... Solved the w/o problem and added sdcard.fth
0.pi@Acer~/cforth$

Then I edited the files and added them as told.
$ git add src/app/esp32-extra/app.fth
$ git add src/app/esp32-extra/extend.c
$ git add src/app/esp32-extra/files.fth
$ git add src/app/esp32-extra/tools/extra.fth
$ git add src/app/esp32-extra/sdcard.fth
$ git add  build/esp32-extra/sdk_build/main/interface.c
$ git rebase --continue

Then the following failed:
$ git push https://github.com/MitchBradley/cforth HEAD:master
remote: Permission to MitchBradley/cforth.git denied to Jos-Ven.
fatal: unable to access 'https://github.com/MitchBradley/cforth/': The requested URL returned error: 403

OK logical.

If I use my own site than that could be confusing for my own 23 cloners.
This whole process of rebasing is very time consuming for both of us. 
Is the following possible: 
1.Throw away or roll back the divergent branches
2.Then I hope I could Open a new pull request from the menu Contribute

Other simple solutions are of course welcome.
Thanks in advance.

Kind regards, Jos

Jos-Ven committed:
Merge branch 'master' into master

Jos-Ven commented:
Perhaps solved by now?

Commit:
Minor correction

quozl commented:
Yes, GitHub has removed the conflict advice.

The problem you raised of discomfort to your clones is why it is more common to make pull requests from a branch other than master.  That way whatever you have to do to change code to be accepted into Mitch's repository won't affect your master branch and anybody relying upon it.

Rebasing becomes quicker if it is something you are aware of before making changes, if you rebase regularly, and if you learn how to do it through practice.

For this pull request, the conflicts and need to rebase was triggered by your work last year on `f>s` in 353a969f07027361f44b3ff9c5b7059be5780274 that came from https://github.com/MitchBradley/cforth/pull/140 and was merged.

In other projects, the need to rebase arises from work by others.

Given the low rate of change in cforth repositories, rebasing is not something you will encounter often.

Once this pull request merges, make sure you pull from Mitch's master branch so that the next changes you make are less likely to show conflicts.

Reviewing your aggregate change;
* changes to src/app/esp32-extra files, as I don't use them (I use platformio build) I'm not concerned with the actual changes, but as I browsed through them I noticed the different style; I would prefer to retain the same stack effect comment style of two spaces before the `(`, and the two spaces before `if` is odd to me too,
* changes to src/cforth/floatops.fth `f>d` reviewed, the indentation is odd, and I've not tested it,
* changes to src/cforth/forth.c `cstr` consumes an extra 80 bytes of stack, and this is because src/app/esp32-extra/extend.c has grown, makes me wonder if there's a better way to handle this growth.

Jos-Ven commented:
> Yes, GitHub has removed the conflict advice.
> 
> The problem you raised of discomfort to your clones is why it is more common to make pull requests from a branch other than master. That way whatever you have to do to change code to be accepted into Mitch's repository won't affect your master branch and anybody relying upon it.
> 
> Rebasing becomes quicker if it is something you are aware of before making changes, if you rebase regularly, and if you learn how to do it through practice.
> 
> For this pull request, the conflicts and need to rebase was triggered by your work last year on `f>s` in [353a969](https://github.com/MitchBradley/cforth/commit/353a969f07027361f44b3ff9c5b7059be5780274) that came from #140 and was merged.
> 
> In other projects, the need to rebase arises from work by others.
> 
> Given the low rate of change in cforth repositories, rebasing is not something you will encounter often.
> 
> Once this pull request merges, make sure you pull from Mitch's master branch so that the next changes you make are less likely to show conflicts.
> 
> Reviewing your aggregate change;
> 
> * changes to src/app/esp32-extra files, as I don't use them (I use platformio build) I'm not concerned with the actual changes, but as I browsed through them I noticed the different style; I would prefer to retain the same stack effect comment style of two spaces before the `(`, and the two spaces before `if` is odd to me too,
> * changes to src/cforth/floatops.fth `f>d` reviewed, the indentation is odd, and I've not tested it,
> * changes to src/cforth/forth.c `cstr` consumes an extra 80 bytes of stack, and this is because src/app/esp32-extra/extend.c has grown, makes me wonder if there's a better way to handle this growth.

Hi, Thank you for your reviewing my changes and your explanations about rebasing.
The problem with rebasing for me is that at one time my local master rolled back all my updates. Luckily there was a backup before rebasing.
The 2 spaces before if were added when I found them more than 249 times in other sources in the cforth directory tree. Your comment makes me wonder if there is some kind of document with recommended guide lines about the layout for sources? Still getting rid of my old style.
The platformio is certainly interesting. I use 12 Esp32 systems in a smart home project. That is very a dynamic environment and involves a lot flashing and updates to the file system of the Esp32. So speed has a very high priority here.
Updating without platformio and using updates through a wifi connection (partly UDP) seems to be the fastest way.
Still looking for faster ways...

Kind regards Jos

quozl commented:
Agree with you that when you have a master branch used by others, rebasing is not helpful.  This is another reason why making pull requests from a branch other than master has become a convention.  GitHub supports this convention by condtitionally granting the target repository developers permission to update the branch connected to a pull request.

When you're not confident you know what will happen with a git command, it is very helpful to try it out first in a backup, or to have a backup you can restore.

https://wiki.laptop.org/go/Forth_Lesson_2 is a helpful write-up of Mitch's style that I've adopted.  I've also learned the style by repeated exposure.  Coding for me is mostly a reading activity, with occasional writing, so writing is best done in the same format that I've been reading.  I like to avoid style changes on a per-file basis.

Yes, I'm familiar with network deployment to esp32.  PlatformIO does support this to an extent.  I'm also experimenting with other hardware, as pricing and availability changes.  My latest experiments are with the Raspberry Pi Pico W modules, where I've tried both MicroPython and CircuitPython, and my preference at the moment is the web workflow of CircuitPython.

Jos-Ven commented:
Hi Mitch,
Update #142 reached it's end when extra.fth was updated a few days ago.
I think I made the changes you asked. Git forced me to upload a newer version of extra.fth due to a divergent branch.
My old formatting style came from the past and was combined with my own opinions. There I grouped small tasks inside a word by adding new lines and extra spaces in it. Then you could immediately see if you forgot something.
Refactoring is not always an option.
Reasons: Dictionary overload, possible performance loss and I am a bad poet.
Thanks in advance, Kind regards, Jos
